icicle/backend/cpu/include/ntt_task.h

```diff
@@ -0,0 +1,1480 @@
+    switch (subntt_size_log) {
+    case 3:
+      ntt8win();
+      break;
+    case 4:
+      ntt16win();
+      break;
+    case 5:
+      ntt32win();
+      break;
+    default:
+      reorder_by_bit_reverse();
+      hierarchy_0_dit_ntt(); // R --> N
+      break;
+    }
+
+    if (
+      ntt_task_coordinates->hierarchy_0_layer_idx != 2 &&
+      ntt_data->ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_coordinates->hierarchy_1_layer_idx]
+                                                        [ntt_task_coordinates->hierarchy_0_layer_idx + 1] != 0) {
+      refactor_output_hierarchy_0();
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  /**
+   * @brief Performs an optimized NTT transformation using a Winograd approach for size 8.
+   *
+   * This function applies the Number Theoretic Transform (NTT) on a sub-NTT of size 8 using
+   * a specialized Winograd algorithm. It utilizes precomputed twiddle factors tailored for
+   * the Winograd algorithm to enhance performance. The function handles both forward and inverse
+   * transformations based on the NTT direction specified in `ntt_data`.
+   *
+   * @return void
+   */
+  template <typename S, typename E>
+  void NttTask<S, E>::ntt8win() // N --> N
+  {
+    uint32_t offset = ntt_data->config.columns_batch ? ntt_data->config.batch_size : 1;
+    E* subntt_elements =
+      ntt_data->elements +
+      offset * (ntt_task_coordinates->hierarchy_1_subntt_idx
+                << ntt_data->ntt_sub_logn
+                     .hierarchy_1_layers_sub_logn[ntt_task_coordinates->hierarchy_1_layer_idx]); // input + subntt_idx *
+                                                                                                 // subntt_size
+    const S* twiddles = ntt_data->direction == NTTDir::kForward
+                          ? CpuNttDomain<S>::s_ntt_domain.get_winograd8_twiddles()
+                          : CpuNttDomain<S>::s_ntt_domain.get_winograd8_twiddles_inv();
+
+    E T;
+    std::vector<uint32_t> index_in_mem(8);
+    uint32_t stride = ntt_data->config.columns_batch ? ntt_data->config.batch_size : 1;
+    for (uint32_t i = 0; i < 8; i++) {
+      index_in_mem[i] = stride * idx_in_mem(ntt_task_coordinates, i);
+    }
+    for (uint32_t batch = 0; batch < ntt_data->config.batch_size; ++batch) {
+      E* current_elements = ntt_data->config.columns_batch ? subntt_elements + batch
+                                                           : subntt_elements + batch * (ntt_data->ntt_sub_logn.size);
+
+      if (
+        ntt_task_coordinates->hierarchy_1_layer_idx == 0 && ntt_task_coordinates->hierarchy_0_layer_idx == 0 &&
+        ntt_data->config.coset_gen != S::one() && ntt_data->direction == NTTDir::kForward) {
+        apply_coset_multiplication(current_elements, index_in_mem, CpuNttDomain<S>::s_ntt_domain.get_twiddles());
+      }
+
+      T = current_elements[index_in_mem[3]] - current_elements[index_in_mem[7]];
```

Comment:
not sure if we tryed this but, can idx_in_mem be an array of pointers(E*) pointing to the element in current_elements?